feeds/wlan-ap/opensync/patches/17-dpp-sta-events.patch

```diff
@@ -0,0 +1,1489 @@
++}
++
++static void mlme_ev_assoc(struct nl_msg *msg, struct nlattr **tb)
++{
++    // check ssid and mac
++    struct nlattr *ssid = tb[NL80211_ATTR_SSID];
++    struct nlattr *mac = tb[NL80211_ATTR_MAC];
++    if(!(ssid && mac)) return;
++
++    // add to list
++    target_event_record_t *event_entry = target_event_record_alloc();
++    event_entry->record.event_type = ET_CLIENT_ASSOC;
++    // client_auth event pointer
++    dpp_event_record_assoc_t *cap = &event_entry->record.client_assoc_event;
++    memset(cap, 0, sizeof(dpp_event_record_assoc_t));
++    // set data
++    memcpy(cap->sta_mac, nla_data(mac), ETH_ALEN);
++    memcpy(cap->ssid, nla_data(ssid), nla_len(ssid));
++    cap->session_id = get_timestamp();
++    // add to global list
++    ds_dlist_insert_tail(&g_rt_events.event_list, event_entry);
++}
++
++static void mlme_ev_disconnect(struct nl_msg *msg, struct nlattr **tb)
++{
++    // check ssid and mac
++    struct nlattr *ssid = tb[NL80211_ATTR_SSID];
++    struct nlattr *mac = tb[NL80211_ATTR_MAC];
++    if(!(ssid && mac)) return;
++
++    // add to list
++    target_event_record_t *event_entry = target_event_record_alloc();
++    event_entry->record.event_type = ET_CLIENT_DISCONNECT;
++    // client_auth event pointer
++    dpp_event_record_disconnect_t *cap = &event_entry->record.client_disconnect_event;
++    memset(cap, 0, sizeof(dpp_event_record_disconnect_t));
++    // set data
++    memcpy(cap->sta_mac, nla_data(mac), ETH_ALEN);
++    memcpy(cap->ssid, nla_data(ssid), nla_len(ssid));
++    cap->session_id = get_timestamp();
++    // add to global list
++    ds_dlist_insert_tail(&g_rt_events.event_list, event_entry);
++}
++
++
++
++static void mlme_ev_recv(struct nl_msg *msg, struct nlattr **tb)
++{
++    //const uint8_t *data;
++    size_t len;
++    struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
++    struct nlattr *frame = tb[NL80211_ATTR_FRAME];
++    struct nlattr *addr = tb[NL80211_ATTR_MAC];
++
++    // frame should be present
++    if(!(frame && addr)) return;
++
++    //data = nla_data(frame);
++    len = nla_len(frame);
++
++    // length check
++    if (len < 4 + 2 * ETH_ALEN) return;
++
++    switch(gnlh->cmd){
++    case NL80211_CMD_AUTHENTICATE:
++            mlme_ev_auth(msg, tb);
++            break;
++    case NL80211_CMD_ASSOCIATE:
++            mlme_ev_assoc(msg, tb);
++            break;
++    case NL80211_CMD_DEAUTHENTICATE:
++    case NL80211_CMD_DISASSOCIATE:
++    case NL80211_CMD_UNPROT_DEAUTHENTICATE:
++    case NL80211_CMD_UNPROT_DISASSOCIATE:
++            mlme_ev_disconnect(msg, tb);
++            break;
++
++    }
++
++}
++
+ static int nl80211_recv(struct nl_msg *msg, void *arg)
+ {
+ 	struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
+@@ -433,6 +537,23 @@ static int nl80211_recv(struct nl_msg *m
+ 	case NL80211_CMD_GET_WIPHY:
+ 		nl80211_add_phy(tb, phyname);
+ 		break;
++
++        // connect
++        case NL80211_CMD_ASSOCIATE:
```

Comment:
tab vs space

Reply:
The eternal battle of tabs vs spaces. Again, my Vim setup.